Source/API/CBLReplication.m

```diff
@@ -403,46 +409,84 @@ - (void) updateStatus: (CBLReplicationStatus)status
 
 #pragma mark - PUSH REPLICATION ONLY:
 
+
+- (CBL_ReplicatorSettings*) replicatorSettings {
+    CBLStatus status;
+    CBL_ReplicatorSettings* settings;
+    settings = [_database.manager replicatorSettingsWithProperties: self.properties
+                                                        toDatabase: nil
+                                                            status: &status];
+    if (!settings)
+        Warn(@"Error parsing replicator settings : %@", CBLStatusToNSError(status));
+    return settings;
+}
+
+
+- (SInt64) lastSequencePushed {
+    if (_pull)
+        return -1;
+    if (_lastSequencePushed <= 0) {
+        // If running replicator hasn't updated yet, fetch the checkpointed last sequence:
+        CBL_ReplicatorSettings* settings = self.replicatorSettings;
+        if (!settings)
+            return -1;
+        _lastSequencePushed = [[_database lastSequenceForReplicator: settings] longLongValue];
+    }
+    return _lastSequencePushed;
+}
+
+
 - (NSSet*) pendingDocumentIDs {
-    if (_pull || (_started && _pendingDocIDs))
-        return _pendingDocIDs;
-
-    _pendingDocIDs = [_database.manager.backgroundServer
-                      waitForDatabaseNamed: _database.name to: ^id(CBLDatabase* bgDb)
-    {
-        CBLStatus status;
-        CBL_ReplicatorSettings* settings =
-            [bgDb.manager replicatorSettingsWithProperties: self.properties
-                                                toDatabase: nil
-                                                    status: &status];
-        if (!settings) {
-            Warn(@"Error parsing replicator settings : %@", CBLStatusToNSError(status));
-            return nil;
-        }
+    if (_pull)
+        return nil;
+    if (_pendingDocIDs) {
+        if (_pendingDocIDsSequence == _database.lastSequenceNumber)
+            return _pendingDocIDs;      // Still valid
+        _pendingDocIDs = nil;
+    }
 
-        NSString* lastSequence = _bg_replicator.lastSequence;
-        if (!lastSequence)
-            lastSequence = [bgDb lastSequenceForReplicator: settings];
-
-        NSError* error;
-        CBL_RevisionList* revs = [bgDb unpushedRevisionsSince: lastSequence
-                                                       filter: settings.filterBlock
-                                                       params: settings.filterParameters
-                                                        error: &error];
-        if (revs)
-            return [NSSet setWithArray: revs.allDocIDs];
-        else
-            Warn(@"Error getting unpushed revisions : %@", error);
+    CBL_ReplicatorSettings* settings = self.replicatorSettings;
+    if (!settings)
         return nil;
-    }];
+    SequenceNumber lastSequence = self.lastSequencePushed;
+    if (lastSequence < 0)
+        return nil;
+
+    _pendingDocIDsSequence = _database.lastSequenceNumber;
```

Comment:
You're right, the assignment should go there. But reading `_database.lastSequenceNumber` should still be before, to avoid skipping any sequences. I'll make the change and update the PR.